devices/SPS30/SHDLC/SHDLCProtocol.cs

```diff
@@ -0,0 +1,177 @@
+    /// Information about SHDLC can be found at https://sensirion.github.io/python-shdlc-driver/shdlc.html
+    /// </remarks>
+    public class SHDLCProtocol
+    {
+        /// <summary>
+        /// The SerialPort is exposed as public property, but should only be used in very specific circumstances. One example is the Sensirion SPS30, for which a raw "0xFF" must be sent to wake the device from its sleep.
+        /// </summary>
+        public SerialPort Serial { get; private set; }
+
+        /// <summary>
+        /// Wraps a serial port to be used for the SHDLC protocol that some Sensirion devices use.
+        /// </summary>
+        /// <param name="serialPort">The serial port to be used.</param>
+        /// <param name="timeoutInMillis">Timeout to be used on the serial port and as communication timeout for commands.</param>
+        public SHDLCProtocol(SerialPort serialPort, int timeoutInMillis = 10000)
+        {
+            Serial = serialPort;
+            Serial.ReadTimeout = timeoutInMillis;
+            Serial.WriteTimeout = timeoutInMillis;
+        }
+
+        /// <summary>
+        /// Execute a command on the SHDLC device. Its documentation should describe the address, commands and what data to send/receive.
+        /// </summary>
+        public byte[] Execute(byte devaddr, byte command, byte[] data)
+        {
+            // Make sure the serial port is opened
+            if (!Serial.IsOpen)
+                Serial.Open();
+
+            // Generate and send our frame
+            var frameBuffer = GenerateMOSIFrame(devaddr, command, data);
+            Serial.Write(frameBuffer, 0, frameBuffer.Length);
+
+            // Read reply and parse response
+            return ReadMOSIFrameContent(devaddr, command);
+        }
+
+        private byte[] GenerateMOSIFrame(byte devaddr, byte command, byte[] data)
+        {
+            // Calculate checksum
+            byte chk = 0;
+            chk += devaddr;
+            chk += command;
+            chk += (byte)data.Length;
+            foreach (var b in data)
+                chk += b;
+            chk = (byte)(0xff ^ chk);
+
+            // Create frame
+            using var frame = new MemoryStream();
+            frame.WriteByte(0x7e);
+            frame.WriteByteStuffed(devaddr);
+            frame.WriteByteStuffed(command);
+            frame.WriteByteStuffed((byte)data.Length);
+            frame.WriteStuffed(data, 0, data.Length);
+            frame.WriteByteStuffed(chk);
+            frame.WriteByte(0x7e);
+
+            return frame.ToArray();
+        }
+
+        /// <summary>
+        /// To allow for garbage, corruption or old responses in the serial buffer, we'll keep reading until there's no more data. Then we return the last valid frame, or the most recent error.
+        /// </summary>
+        private byte[] ReadMOSIFrameContent(byte expectedDevAddr, byte expectedCommand)
```

Comment:
`ReadMosiFrameContent`

and for consistency with the Generate function, either remove Content either add it to the other function?

Reply:
This is intentional. When we generate a frame, we generate the whole thing including header bytes and content, which is then passed to the `SerialPort` to write. The `ReadMosiFrameContent` reads the response, validates the response, and only returns the actually requested data (which could also be an empty byte array). I.e., actually the content of the reply.

While it doesn't seem consistent, it does describe what its intent is.